src/rpc/mining.cpp

```diff
@@ -733,15 +734,20 @@ static UniValue submitblock(const JSONRPCRequest& request)
     bool new_block;
     submitblock_StateCatcher sc(block.GetHash());
     RegisterValidationInterface(&sc);
-    bool accepted = ProcessNewBlock(Params(), blockptr, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
+    CValidationState dos_state;
+    bool accepted = ProcessNewBlock(Params(), blockptr, dos_state, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
     UnregisterValidationInterface(&sc);
     if (!new_block && accepted) {
         return "duplicate";
     }
-    if (!sc.found) {
+    if (!sc.found && dos_state.IsValid()) {
         return "inconclusive";
     }
-    return BIP22ValidationResult(sc.state);
+    if (dos_state.IsValid()) {
+        return BIP22ValidationResult(sc.state);
+    } else {
+        return BIP22ValidationResult(dos_state);
+    }
```

Comment:
Consider the less verbose version:

```c++
return BIP22ValidationResult(dos_state.IsValid() ? sc.state : dos_state);
```

Or use a guard clause to be consistent with the previous `if`:

```c++
if (!dos_state.IsValid()) {
    return BIP22ValidationResult(dos_state);
}

return BIP22ValidationResult(sc.state);
```